### ISSUE TITLE
Database Update Failure Modes & Retry Duration

### ISSUE THREAD

**Issue by cpu** [google/safebrowsing#44]
Hi there!

I've been working through integrating this library into [Boulder](https://github.com/letsencrypt/boulder) to replace our usage of the legacy v3 API for checking domain names prior to issuing a certificate. I thought it would be valuable to provide some feedback of things we've noticed so far. There are some rough edges that perhaps you have solutions for! 

In the process of integration testing with a mock server I've noticed the same behaviour as @smoya in https://github.com/google/safebrowsing/issues/37 - If the initial `/v4/threatListUpdates:fetch` call done after constructing a safebrowsing client results in error the library is unable to do any lookups until the situation is rectified (with the `DefaultUpdatePeriod` this is no sooner than 30 minutes later). This could also be caused by a failed update on refresh requests.

It seems like there should be a separate "knob" to twist for a retry period when an error occurs. If we change `DefaultUpdatePeriod` to attempt to recover more quickly we will also produce unnecessarily frequent lookups after the error is resolved.

It's also not ideal that the library isn't able to communicate when it is ready to provide lookups with a healthy database. We can work around this by performing our own polling on the client's `Status` function looking for a `nil` return value for the `error` return parameter but ideally there would be a way
for the integrating system to block on, or be notified of, when the database is healthy and the client is capable of doing lookups.

Thoughts? Suggestions? 

Thanks for all your work on this project!

**Comment by cpu** [comment]
@colonelxc Are you the right person to bump about this?

**Comment by alexwoz** [comment]
Hi @cpu, thanks for your message, and apologies for the latency in responding. I'm supportive of your suggestions :)

We actually have a [back-off spec](https://developers.google.com/safe-browsing/v4/request-frequency#back-off-mode) in our public docs that this client doesn't currently abide by. Since backend errors are rare, implementing this policy should help in getting the client back to a healthy state in a more timely fashion.

+1 for a blocking subscription channel for notifying once the client is ready.

I'm not sure how quickly we'd be able to commit these changes, but we'd be happy to review any code if you'd like to get these in sooner.

Thanks!
Alex

**Comment by cpu** [comment]
> Hi @cpu, thanks for your message, and apologies for the latency in responding. I'm supportive of your suggestions :)

Not a problem! Thanks for replying :-)

> We actually have a back-off spec in our public docs that this client doesn't currently abide by. Since backend errors are rare, implementing this policy should help in getting the client back to a healthy state in a more timely fashion.

This seems sensible and would definitely be nicer than waiting for the next scheduled database refresh.

> I'm not sure how quickly we'd be able to commit these changes, but we'd be happy to review any code if you'd like to get these in sooner.

Understood. I think for the short term we're going to stick with our existing V3 implementation and wait for the improvements we need in the V4 client library. It's unlikely I'll be able to find time in the next few sprints for a PR from our side but if we decide adopting V4 is a priority I'll revisit our engagement too.

Thanks!

**Comment by smoya** [comment]
It would be nice to have this implemented asap. 

Thanks for your message @alexwoz

**Comment by cpu** [comment]
One half of the issues raised (the backoff behaviour) were fixed with #49. The other half will be addressed by #52 and I believe this issue should be closed after it is merged.

**Comment by colonelxc** [comment]
I believe this issue is fully resolved now. Thanks for being patient.
